compiler-core/src/inline.rs

```diff
@@ -0,0 +1,1786 @@
+    /// After beta reduction:
+    /// ```gleam
+    /// let x = Ok(10)
+    /// {
+    ///   let result = x
+    ///   let function = fn(x) {
+    ///     let y = x + 4
+    ///     int.to_string(y)
+    ///   }
+    ///   case result {
+    ///     Ok(value) -> Ok(function(value))
+    ///     Error(error) -> Error(error)
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// And finally, after the final optimising pass, where this inlining process
+    /// is repeated:
+    /// ```gleam
+    /// let x = Ok(10)
+    /// case x {
+    ///   Ok(value) -> Ok({
+    ///     let y = x + 4
+    ///     int.to_string(y)
+    ///   })
+    ///   Error(error) -> Error(error)
+    /// }
+    /// ```
+    ///
+    fn call(
+        &mut self,
+        location: SrcSpan,
+        type_: Arc<Type>,
+        function: Box<TypedExpr>,
+        arguments: Vec<TypedCallArg>,
+    ) -> TypedExpr {
+        let arguments = arguments
+            .into_iter()
+            .map(
+                |TypedCallArg {
+                     label,
+                     location,
+                     value,
+                     implicit,
+                 }| TypedCallArg {
+                    label,
+                    location,
+                    value: self.expression(value),
+                    implicit,
+                },
+            )
+            .collect();
+
+        // First, we traverse the left-hand side of this call. If this is called
+        // inside another inlined function, this could potentially inline an
+        // argument, allowing further inlining.
+        let function = self.expression(*function);
+
+        // If the left-hand side is in a block for some reason, for example
+        // `{ fn(x) { x + 1 } }(10)`, we still want to be able to inline it.
+        let function = expand_block(function);
+
+        let function = match function {
+            TypedExpr::Var {
+                ref constructor,
+                ref name,
+                ..
+            } => match &constructor.variant {
+                ValueConstructorVariant::ModuleFn { module, .. } => {
+                    // If the function is in the list of inlinable functions in
+                    // the module it belongs to, we can inline it!
+                    if let Some(function) = self
+                        .modules
+                        .get(module)
+                        .and_then(|module| module.inline_functions.get(name))
+                    {
+                        // First, we do the actual inlining, by converting it to
+                        // an anonymous function.
+                        let TypedExpr::Fn {
+                            args: parameters,
+                            body,
+                            ..
+                        } = function.to_anonymous_function()
+                        else {
+                            unreachable!("to_anonymous_function always returns TypedExpr::fn");
+                        };
```

Comment:
Don't panic when it's possible to write code where the invariants are enforced using the type system! Instead of returning a typed expr which _might_ be a ::fn, return the arguments and the body, as that's the only data that is actually needed.